homeassistant/components/bluesound/media_player.py

```diff
@@ -884,121 +702,105 @@ async def async_unjoin(self):
+            if preset.name == source:
+                url = preset.url
 
-        selected_source = items[0]
-        url = f"Play?url={selected_source['url']}&preset_id&image={selected_source['image']}"
-
-        if selected_source.get("is_raw_url"):
-            url = selected_source["url"]
-
-        return await self.send_bluesound_command(url)
+        async with timeout(BluesoundPlayer.COMMAND_TIMEOUT):
+            await self._player.play_url(url)
 
     async def async_clear_playlist(self) -> None:
         """Clear players playlist."""
         if self.is_grouped and not self.is_master:
             return
 
-        return await self.send_bluesound_command("Clear")
+        async with timeout(BluesoundPlayer.COMMAND_TIMEOUT):
+            await self._player.clear()
 
     async def async_media_next_track(self) -> None:
         """Send media_next command to media player."""
         if self.is_grouped and not self.is_master:
             return
 
-        cmd = "Skip"
-        if self._status and "actions" in self._status:
-            for action in self._status["actions"]["action"]:
-                if "@name" in action and "@url" in action and action["@name"] == "skip":
-                    cmd = action["@url"]
-
-        return await self.send_bluesound_command(cmd)
+        async with timeout(BluesoundPlayer.COMMAND_TIMEOUT):
+            await self._player.skip()
 
     async def async_media_previous_track(self) -> None:
         """Send media_previous command to media player."""
         if self.is_grouped and not self.is_master:
             return
 
-        cmd = "Back"
-        if self._status and "actions" in self._status:
-            for action in self._status["actions"]["action"]:
-                if "@name" in action and "@url" in action and action["@name"] == "back":
-                    cmd = action["@url"]
-
-        return await self.send_bluesound_command(cmd)
+        async with timeout(BluesoundPlayer.COMMAND_TIMEOUT):
+            await self._player.back()
 
     async def async_media_play(self) -> None:
         """Send media_play command to media player."""
         if self.is_grouped and not self.is_master:
             return
 
-        return await self.send_bluesound_command("Play")
+        async with timeout(BluesoundPlayer.COMMAND_TIMEOUT):
+            await self._player.play()
 
     async def async_media_pause(self) -> None:
         """Send media_pause command to media player."""
         if self.is_grouped and not self.is_master:
             return
 
-        return await self.send_bluesound_command("Pause")
+        async with timeout(BluesoundPlayer.COMMAND_TIMEOUT):
+            await self._player.pause()
 
     async def async_media_stop(self) -> None:
         """Send stop command."""
         if self.is_grouped and not self.is_master:
             return
 
-        return await self.send_bluesound_command("Pause")
+        async with timeout(BluesoundPlayer.COMMAND_TIMEOUT):
```

Comment:
Why don't we put this timeout in the library instead?

Reply:
The old code did not use the timeout functionality of `aiohttp` and used `async with timeout(...)` instead and i kept it that way.

Reply:
So it essentially does the same?

Reply:
Yes it should be equivalent. I had a quick look at `aiohttp` and it seems to use `asyncio.timeout_at` internally.

Reply:
Honestly, if we can already make the code cleaner (by moving connection specific things to the library, like timeouts), go for it

Reply:
It’s always better to put the time out as far down in the code as possible so the least amount of has to worry about a cancellation.

In this case, it’s much much better for it to be done in aiohttp

Reply:
I moved the timeouts into the library. Most of the explicit timeouts have been replaced by a default timeout for all requests.